src/Stratis.Bitcoin.Features.MemoryPool.Tests/MempoolValidatorTest.cs

```diff
@@ -411,71 +412,381 @@ public async Task AcceptToMemoryPool_WithSegWitValidTxns_IsSuccessfullAsync()
+            ITestChainContext context = await TestChainFactory.CreateAsync(KnownNetworks.RegTest, minerSecret.PubKey.Hash.ScriptPubKey, dataDir);
+            IMempoolValidator validator = context.MempoolValidator;
+            Assert.NotNull(validator);
+
+            var destSecret = new BitcoinSecret(new Key(), KnownNetworks.RegTest);
+            var tx = new Transaction();
+
+            // Create a transaction that looks like a coinbase. But the consensus rules that apply to coinbases will reject it anyway.
+            // We need two forms of the test because submitting an incorrectly constructed coinbase triggers a consensus rule instead
+            // of the coinbase being rejected by the mempool for being a coinbase.
+            // TODO: Instead of creating multiple versions of all such tests we should perhaps find a way of testing the applicable rules in the mempool simultaneously.
+            tx.AddInput(new TxIn(new OutPoint(new uint256(0), uint.MaxValue), PayToPubkeyHashTemplate.Instance.GenerateScriptPubKey(minerSecret.PubKey)));
+            tx.AddOutput(new TxOut(new Money(Money.CENT * 11), destSecret.PubKeyHash));
+            tx.Sign(KnownNetworks.RegTest, minerSecret, false);
+            Assert.True(tx.IsCoinBase);
+
+            var state = new MempoolValidationState(false);
+            bool isSuccess = await validator.AcceptToMemoryPool(state, tx);
+
+            // Tests PreMempoolChecks invokes CheckPowTransactionRule and is enforced for mempool transactions.
+            Assert.False(isSuccess, "Coinbase with incorrect size should not be accepted to mempool.");
         }
 
-        [Fact(Skip = "Not implemented yet.")]
-        public void AcceptToMemoryPool_TxIsNonStandardTransactionSizeInvalid_ReturnsFalse()
+        [Fact]
+        public async void AcceptToMemoryPool_TxIsCoinstake_ReturnsFalseAsync()
         {
-            // TODO:Test the cases in PreMempoolChecks CheckStandardTransaction
-            // - Check transaction size
+            string dataDir = GetTestDirectoryPath(this);
+
+            var minerSecret = new BitcoinSecret(new Key(), KnownNetworks.RegTest);
+            ITestChainContext context = await TestChainFactory.CreateAsync(KnownNetworks.RegTest, minerSecret.PubKey.Hash.ScriptPubKey, dataDir);
+            IMempoolValidator validator = context.MempoolValidator;
+            Assert.NotNull(validator);
+
+            var destSecret = new BitcoinSecret(new Key(), KnownNetworks.RegTest);
+            var tx = new Transaction();
+
+            // Create a transaction that looks like a coinstake.
+            tx.AddInput(new TxIn()
+            {
+                PrevOut = new OutPoint(new uint256(15), 1),
+                ScriptSig = new Script()
+            });
+            tx.AddOutput(new TxOut(Money.Zero, (IDestination)null));
+            tx.AddOutput(new TxOut(Money.Zero, (IDestination)null));
+            Assert.True(tx.IsCoinStake);
```

Comment:
this is very odd if you did above `var tx = new Transaction();`

Reply:
I think `IsCoinStake` checks the structure of the transaction rather than its type, perhaps we should add a more strict check.

Reply:
hmm yeah correct if you dont serialize it, but its fine like this (we just need to remember to use the factory)